src/shogun/regression/KernelRidgeRegression.cpp

```diff
@@ -66,21 +65,10 @@ bool KernelRidgeRegression::solve_krr_system()
 
 bool KernelRidgeRegression::train_machine(std::shared_ptr<Features >data)
 {
-	require(m_labels, "No labels set");
-
-	if (data)
-	{
-		if (m_labels->get_num_labels() != data->get_num_vectors())
-			error("Number of training vectors does not match number of labels");
-		kernel->init(data, data);
-	}
-	ASSERT(kernel && kernel->has_features())
-
-	if (m_labels->get_num_labels() != kernel->get_num_vec_rhs())
-	{
-		error("Number of labels does not match number of kernel"
-			" columns (num_labels={} cols={}", m_labels->get_num_labels(), kernel->get_num_vec_rhs());
-	}
+	require(m_labels->get_num_labels() == data->get_num_vectors(), 
```

Comment:
Wait, wouldnt labels be provided as a parameter here? Or has the member been set elsewhere? I think in both parametric and nonparametric, the train method should have feats,labels as parameters. In nonparametric these are then stored locally

Reply:
yes, the m_labels have been set in `train`.
https://github.com/shogun-toolbox/shogun/blob/3d99460a74f40649645488f095a75ea21be38df6/src/shogun/machine/Machine.cpp#L74-L77

Reply:
The idea is that when the refactor is done we can remove m_labels from Machine, and then we can remove the set_labels in train

Reply:
but then we need to touch everything *again* ... imo the labels should be passed to `train_machine` (also to be consistent with feature being passed). We should never (in train) rely on labels/features being retrieved from members.

Reply:
but then we need to change everything in one go no? I think the idea was to do these small PRs for npmachine, and when they are done we can have a big PR where we just replace the function signatures?
The idea is that then we can also uncomment this
https://github.com/shogun-toolbox/shogun/blob/3d99460a74f40649645488f095a75ea21be38df6/src/shogun/machine/NonParametricMachine.h#L22-L24

Reply:
i see. so signatures come later. sorry about my moans then :)
It is just all these files need to be touched again. But I guess it will be considerably easier once these changes here are done

Reply:
I think at the start we looked at also changing the signatures, but then we also thought about making things like `train_machine` pure virtual, so then we will need to change all that too. Also could make `apply` `const`. So doing all that when the labels are separated more from the implementation is imo simpler, but I also don't have to write the code :D